### PR TITLE
Time Lock Upgradability

### PR DESCRIPTION
**Pull request by felix2feng**
- Moved TimeLockUpgrade into its own Library
- TimeLockUpgrade is used in `CoreInternal` and `Authorizable`

### CODE REVIEW

File: test/contracts/core/extensions/coreTimeLockUpgrade.spec.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+          { from: subjectCaller },
+        );
+      });
+
+      it('should register the timeLockedUpgrade with the transaction timestamp', async () => {
+        const txHash = await subject();
+        const { blockHash, input } = await web3.eth.getTransaction(txHash);
+        const { timestamp } = await web3.eth.getBlock(blockHash);
+
+        const upgradeHash = web3.utils.soliditySha3(input);
```
Comment (felix2feng) [inline test/contracts/core/extensions/coreTimeLockUpgrade.spec.ts]: Really cool that web3 1.0 now has a pack and hash function

File: test/contracts/core/extensions/coreTimeLockUpgrade.spec.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      beforeEach(async () => {
+        await coreTimeLockUpgradeMock.setTimeLockPeriod.sendTransactionAsync(
+          ONE,
+          { from: subjectCaller },
+        );
+      });
+
+      it('should register the timeLockedUpgrade with the transaction timestamp', async () => {
+        const txHash = await subject();
+        const { blockHash, input } = await web3.eth.getTransaction(txHash);
```
Comment (felix2feng) [inline test/contracts/core/extensions/coreTimeLockUpgrade.spec.ts]: `input` = calldata

File: contracts/core/extensions/CoreInternal.sol [hunk 3]
```diff
@@ -240,7 +246,26 @@ contract CoreInternal is
     )
         external
         onlyOwner
+        timeLockUpgrade
     {
         state.signatureValidator = _signatureValidator;
     }
+
```
Comment (bweick) [inline contracts/core/extensions/CoreInternal.sol]: Why not put this in a constructor instead of having a function that only gets called once?

Comment (felix2feng) [reply contracts/core/extensions/CoreInternal.sol]: It kills 2 birds with 1 stone, which is 1) adding an initial period where the timelock doesn't apply (not setting it has that effect) and 2) allows adding that timelock

Comment (bweick) [reply contracts/core/extensions/CoreInternal.sol]: I see. That makes sense

**Pull request merged**